examples/hdf5.py

```diff
@@ -4,99 +4,177 @@
+of plot updates. Further, a relatively small amount of downsampled data is precomputed and stored
+in the HDF5 file, which is sufficient for navigation of the entire data set without aliasing or lag.
 """
 
-import initExample ## Add path to library (just for examples; you do not need this)
+import initExample  # Add path to library (just for examples; you do not need this)
+
+import os
+import sys
 
-import sys, os
-import numpy as np
 import h5py
+import numpy as np
 import pyqtgraph as pg
 from pyqtgraph.Qt import QtCore, QtGui
 
 pg.mkQApp()
 
-
 plt = pg.plot()
 plt.setWindowTitle('pyqtgraph example: HDF5 big data')
 plt.enableAutoRange(False, False)
 plt.setXRange(0, 500)
 
-class HDF5Plot(pg.PlotCurveItem):
+
+class BigDataPlot(pg.PlotCurveItem):
     def __init__(self, *args, **kwds):
-        self.hdf5 = None
-        self.limit = 10000 # maximum number of samples to be plotted
+        self._plotData = None
         pg.PlotCurveItem.__init__(self, *args, **kwds)
-        
-    def setHDF5(self, data):
-        self.hdf5 = data
-        self.updateHDF5Plot()
-        
+
+    def setPlotData(self, plotData):
+        self._plotData = plotData
+        self.replot()
+
     def viewRangeChanged(self):
-        self.updateHDF5Plot()
-        
-    def updateHDF5Plot(self):
-        if self.hdf5 is None:
+        self.replot(lazy=True)
+
+    def replot(self, lazy=False):
+        if self._plotData is None:
             self.setData([])
             return
-        
+
         vb = self.getViewBox()
         if vb is None:
             return  # no ViewBox yet
-        
-        # Determine what data range must be read from HDF5
-        xrange = vb.viewRange()[0]
-        start = max(0,int(xrange[0])-1)
-        stop = min(len(self.hdf5), int(xrange[1]+2))
-        
-        # Decide by how much we should downsample 
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
+
+        x1, x2 = vb.viewRange()[0]
+        data = self._plotData.sample(x1, x2, lazy=lazy)
+
+        if data is not None:
+            self.setData(*data)  # update the plot
+
+
+class PlotData:
+    def __init__(self, x, y, cache=None):
+        if x is None:
+            x = len(y)
+        elif len(x) != len(y):
+            raise ValueError
+
+        self._x = x
+        self._y = y
+        self._cache = cache
+        self._lastRange = None
+
+    def sample(self, x1, x2, lazy=True, padding=0.3, **kwargs):
+        i1, i2, ds = plotDataRange(self._x, x1, x2, padding=padding, **kwargs)
+
+        if lazy and self._lastRange is not None:
+            i11, i21, ds1 = self._lastRange
+            i12, i22, _ = plotDataRange(self._x, x1, x2, padding=0, **kwargs)
+            if (i11 <= i12 <= i22 <= i21) and (ds1 == ds):
+                return
```

Comment:
This bit is slightly weird: using `None` to indicate the delayed computation.

I wonder if you could do something a little nicer with this.
I need to reread everything first; might take me a bit 😂

Reply:
So in the lazy case where nothing changes, there's no state set, you're just getting the last range and computing the new one yah?

I wonder does this need to be inside the class then?
I guess to encapsulate all the `i` value details.

I wonder could you not just check against the last `x1`, `x2` inputs instead of recomputing the matching `i` value outputs.
If the function is injective you shouldn't need to check the outputs for each new input, just compare the inputs.